networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/rpc/beaconchain/methods/BlobSidecarsByRootMessageHandler.java

```diff
@@ -0,0 +1,218 @@
+        peer.getId(),
+        message.size(),
+        message);
+
+    if (!peer.wantToMakeRequest() || !peer.wantToReceiveBlobSidecars(callback, message.size())) {
+      requestCounter.labels("rate_limited").inc();
+      return;
+    }
+
+    requestCounter.labels("ok").inc();
+    totalBlobSidecarsRequestedCounter.inc(message.size());
+
+    SafeFuture<Void> future = SafeFuture.COMPLETE;
+
+    final UInt64 finalizedEpoch = getFinalizedEpoch();
+
+    for (final BlobIdentifier identifier : message) {
+      final Bytes32 blockRoot = identifier.getBlockRoot();
+      future =
+          future
+              .thenCompose(__ -> validateRequestedBlockRoot(blockRoot, finalizedEpoch))
+              .thenCompose(__ -> retrieveBlobSidecar(identifier))
+              .thenComposeChecked(
+                  maybeSidecar -> {
+                    if (maybeSidecar.isEmpty()) {
+                      throw new ResourceUnavailableException(
+                          String.format(
+                              "Blob sidecar for blob identifier (%s) was not available",
+                              identifier));
+                    }
+                    return callback.respond(maybeSidecar.get());
+                  });
+    }
+
+    future.finish(callback::completeSuccessfully, err -> handleError(callback, err));
+  }
+
+  // MAX_REQUEST_BLOB_SIDECARS * MAX_BLOBS_PER_BLOCK
+  private int calculateMaxRequestSize() {
+    final UInt64 currentEpoch = combinedChainDataClient.getCurrentEpoch();
+    final int maxBlobsPerBlock =
+        SpecConfigDeneb.required(spec.atEpoch(currentEpoch).getConfig()).getMaxBlobsPerBlock();
+    return MAX_REQUEST_BLOB_SIDECARS.times(maxBlobsPerBlock).intValue();
+  }
+
+  private UInt64 getFinalizedEpoch() {
+    return combinedChainDataClient
+        .getFinalizedBlock()
+        .map(SignedBeaconBlock::getSlot)
+        .map(spec::computeEpochAtSlot)
+        .orElse(UInt64.ZERO);
+  }
+
+  /**
+   * Validations:
+   *
+   * <ul>
+   *   <li>A block for the block root is available
+   *   <li>The block root references a block greater than or equal to the minimum_request_epoch
+   * </ul>
+   */
+  private SafeFuture<Void> validateRequestedBlockRoot(
+      final Bytes32 blockRoot, final UInt64 finalizedEpoch) {
+    return combinedChainDataClient
+        .getBlockByBlockRoot(blockRoot)
+        .thenAcceptChecked(
+            maybeBlock -> {
+              if (maybeBlock.isEmpty()) {
+                throw new ResourceUnavailableException(
+                    String.format("Block for block root (%s) couldn't be retrieved", blockRoot));
+              }
+              final SignedBeaconBlock block = maybeBlock.get();
+              final UInt64 requestedEpoch = spec.computeEpochAtSlot(block.getSlot());
+              final UInt64 minimumRequestEpoch = computeMinimumRequestEpoch(finalizedEpoch);
+              if (requestedEpoch.isLessThan(minimumRequestEpoch)) {
+                throw new RpcException(
```

Comment:
From the spec: 
`If any root in the request content references a block earlier than minimum_request_epoch, peers MAY respond with error code 3: ResourceUnavailable or not include the blob in the response.`. 

So, it's **MAY**. Should we be so strict when responding?

Reply:
I don't know. It's good to have diversity, so decided to include it, since we already retrieve the block and it's cheap.